Mark Task Groups as Success/Failure

tquazi opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
closes: #30220

This pull request adds ability to mark task groups as success/failure and have that propagate to the tasks within that task group. This feature is designed to provide a more efficient way to adjust the status of multiple tasks at once, especially for complex workflows with large numbers of tasks.

One of the primary use cases for this feature is when certain tasks in a task group have failed due to unexpected issues. Instead of manually setting the status of each failed task to "failed", users can mark the entire task group as "failed", and the system will automatically propagate this status to all tasks within the group. Similarly, in the case of a manual DAG run triggered through the UI or the API, where certain tasks or sensors may not matter for that run, users can easily mark the entire task group as "success" to avoid having to set each task individually.

I have tested this feature. Please let me know if there are any further changes or improvements that you would like me to make before merging this pull request.

Thank you for your time and consideration.


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

tquazi committed:
Initial commit

tquazi committed:
Initial commit

utkarsharma2 commented on airflow/models/dag.py:
```diff
@@ -1840,6 +1840,7 @@ def set_task_instance_state(
         past: bool = False,
         commit: bool = True,
         session=NEW_SESSION,
+        group_id: str | None = None,
```
@tquazi It would be nice if we can add new param to docstring :)

tquazi replied:
Added group_id to the docstring.

tquazi committed:
Review comments

tquazi committed:
Review comments

bbovenzi reviewed (approved):
Reviewing the UI code and testing locally. This looks good.
Ultimately, I'd like these actions to live in the REST API instead of the webserver.

After this is merged I'll rebase my actions PR: https://github.com/apache/airflow/pull/30373

tquazi committed:
Fixes

tquazi committed:
Fixes

jedcunningham commented on airflow/models/dag.py:
```diff
@@ -1915,9 +1943,10 @@ def set_task_instance_state(
             session=session,
             # Exclude the task itself from being cleared
-            exclude_task_ids={task_id},
+            exclude_task_ids=task_ids,
```
This type isn't right? I'm not sure why this isn't failing with mypy, but doesn't this expect a frozenset?

tquazi replied:
Yes, you are right. It expects a frozenset and in the old code we were passing a set. I have corrected this now.

jedcunningham commented on airflow/models/dag.py:
```diff
@@ -1840,6 +1840,7 @@ def set_task_instance_state(
         past: bool = False,
         commit: bool = True,
         session=NEW_SESSION,
+        group_id: str | None = None,
```
I'm not sure I like the pattern of `set_task_instance_state` acting on a task instance, or a group when you pass this optional group_id (thus, ignoring the task_id completely). I think I'd rather see a `set_task_group_state` instead.

tquazi replied:
I have implemented a new method `set_task_group_state` for this functionality as suggested. This looks much cleaner now.

jedcunningham reviewed (dismissed):
This also needs test coverage.

tquazi committed:
Test Coverage + Review comment changes

1. Added Test Cases for dag.set_task_group_state and view._mark_task_group_state.
2. Added new method dag.set_task_group_state, which can be used when setting the state on a task group.

tquazi committed:
Test Coverage + Review comment changes

1. Added Test Cases for dag.set_task_group_state and view._mark_task_group_state.
2. Added new method dag.set_task_group_state, which can be used when setting the state on a task group.

ashb commented on airflow/models/dag.py:
```diff
@@ -1915,9 +1914,118 @@ def set_task_instance_state(
+        task_group = task_group_dict.get(group_id)
+        if task_group is None:
+            raise ValueError("TaskGroup {group_id} could not be found")
+        tasks_to_set_state = [task for task in task_group.iter_tasks() if isinstance(task, BaseOperator)]
```
What is this is-instance check guarding against? This will mean that mapped tasks aren't cleared (as those don't inherit from BaseOperator). There is `AbstractOperator` which is probably better to use than BaseOperator

I'd also think it would be good to not add the top level import -- all these model files are intertwined enough, and adding imports to top levle can slow down parsing/running tasks in unexpected ways

tquazi replied:
task_group.iter_tasks returns an iterator of AbstractOperator objects.
I am using is-instance to cast AbstractOperator to BaseOperator. 
The set_state method expects List<BaseOperator> as argument and the TypeChecking is failing this piece of code.

https://github.com/apache/airflow/pull/30478/files#diff-62c8e300ee91e0d59f81e0ea5d30834f04db71ae74f2e155a10b51056b00b59bR1991-R2002

We have similar use case in views.py But looks like TypeChecking is not enabled here.
https://github.com/apache/airflow/blob/main/airflow/www/views.py#L2183

Any suggestions to resolve this?

ashb commented on airflow/models/dag.py:
```diff
@@ -1915,9 +1914,118 @@ def set_task_instance_state(
+        # Clear downstream tasks that are in failed/upstream_failed state to resume them.
+        # Flush the session so that the tasks marked success are reflected in the db.
+        session.flush()
+        subdag = self.partial_subset(
```
Please don't call this subdag -- that has a very specific meaning (that of SubDagOperator) which this is not.

tquazi replied:
Changed the name of the variable :)

ashb commented on airflow/www/views.py:
```diff
@@ -2458,6 +2458,38 @@ def _mark_task_instance_state(
+        dag: DAG = get_airflow_app().dag_bag.get_dag(dag_id)
+
+        if not run_id:
+            flash(f"Cannot mark tasks as {state}, seem that DAG {dag_id} has never run", "error")
```
```suggestion
            flash(f"Cannot mark tasks as {state}, as DAG {dag_id} has never run", "error")
```

tquazi replied:
Fixed

ashb commented on airflow/www/views.py:
```diff
@@ -2474,6 +2506,7 @@ def confirm(self):
         dag_run_id = args.get("dag_run_id")
         state = args.get("state")
         origin = get_safe_url(args.get("origin"))
+        group_id = args.get("group_id")
```
I do wonder if we need separate argument for this -- task_ids and group_ids can never overlap, so given an ID we should be able to find it in the DAG and work out if it's a task or a group. We could call the parameter `node_id`

tquazi replied:
We follow similar pattern for ClearTask. https://github.com/apache/airflow/blob/main/airflow/www/views.py#L2152
Let me know if this needs to change as well.

Having a single parameter `node_id` makes more sense, but we will have to add logic to find if it is a task or a group. From the views we already know if it is a group or not, can we send node_id and isGroup parameters?

ashb commented on airflow/models/dag.py:
```diff
@@ -1915,9 +1914,118 @@ def set_task_instance_state(
+        session: Session = NEW_SESSION,
+    ) -> list[TaskInstance]:
+        """
+        Set the state of the TaskGroup to the given state, and clear its downstream tasks that are
+        in failed or upstream_failed state.
```
Does this comment hold true when the state is set to Success?

tquazi replied:
When we set the state of a Task to Success/Failure, the existing logic is to clear downstream tasks that are in failed or up_stream failed state. I am assuming this to be valid for TaskGroup as well. Am I missing something here. Thanks!

tquazi committed:
Changes after rebase

tquazi committed:
Changes after rebase

tquazi committed:
Merge branch 'taskgroup-actions' of https://github.com/tquazi/airflow into taskgroup-actions

tquazi committed:
Warning for Past/Future tasks

Added warning if marking past/future tasks as Success/Failure

tquazi committed:
Merge branch 'main' into taskgroup-actions

tquazi committed:
Merge branch 'main' into taskgroup-actions
